aten/src/ATen/core/Tensor.h

```diff
@@ -165,10 +166,20 @@ class CAFFE2_API Tensor {
   int64_t ndimension() const {
     return dim();
   }
-  bool is_contiguous() const {
-    return impl_->is_contiguous();
+  bool is_contiguous(at::MemoryFormat memory_format=at::MemoryFormat::Any) const {
+    return impl_->is_contiguous(memory_format);
+  }
+  void set_memory_format_tag(at::MemoryFormat memory_format) {
+    impl_->set_memory_format_tag(memory_format);
+  }
+  void reset_memory_format_tag() {
+    impl_->reset_memory_format_tag();
   }
 
+  // This functions completly ignores current data, can be used on empty tensor only
```

Comment:
completely

Reply:
I have no idea what this function does, and thus I have no idea what this comment is saying

Reply:
And, looking through this patch, it is not documented what this method does anywhere.

Additionally, since this may fail, I think we should use `C10_NODISCARD` here

Reply:
If this function totally corrupts the current data, wouldn't it be safer to have something like `at::empty()` support taking layout explicitly (IIRC, that's part of the original spec, right?) and then make most code use that, instead of this internal function? There are a lot of call sites to this, and this seems way too low level for me to feel comfortable letting random people use it.

Reply:
After reading the use sites, I think I have a better idea what this function does, and the semantics of the function gives me hives.

It seems to do two distinct things:

1. If the tensor is contiguous and 4-D, mutably restride it so that channel dimension is physically last but logically in second place. And label it accordingly.
2. If the tensor is not contiguous, check if the strides are consistent with being channel-last, and if they are, label it accordingly.

I can't think of *any* call site that would need both of these functionalities; it seems we statically always know if we are in case (1) or (2). If that's the case, I'd much rather just have two separate functions that do one thing, or another.